Npm publish

Adamik10 opened the pull request:
#### What does this PR do?
- Introduces the ability to create an NPM package out of the repository that can be used for styling and assets in other projects 
- In connection to this, the README.md is also adjusted for documentation of this feature
- Some Github actions have also been altered in connection to this 

#### Where should the reviewer start?
- Just file by file 

#### How can the reviewer see this in action?
-

#### Any background context you want to provide?
The package functionality was conceived to tackle the problem with another repository - ddb-react  where we previously had to copy paste whole css files in order to take advantage of the styling from this repository. 

#### Is this related to processing of personal data?
No.

#### What are the relevant tickets?
N/A

#### Screenshots (if appropriate)
-

#### Questions for the reviewer:
-

kasperg committed:
Publish package to local GitHub npm registry

This allows developers of Node projects to include the design system
as a dependency and use the output from the package in a similar way
as they would use other packages containing static assets.

Forks and branches are a regular part of the development workflow. To
support publishing across these we need to modify the package.json
file of the project accordingly. The json package is added to support
this.Changes are kept in GitHub Actions and not committed to make it
easier to stay up to date with upstream.

Update documentation accordingly.

kasperg committed:
Fix bundle compatiblity for OSX

The -t argument is not supported.

We also switch from + to \; to terminate file commands since that 
also seems to cause problems.

kasperg committed:
Build releases of all branches - not just main

This allows other projects to track continual development which has
not been merged yet.

kasperg committed:
Use fixed version of GitHub auto releases action

Running with latest can be dangerous when you do not know how the
maintainer will handle major release bumps.

Tools like Dependabot should help us keep up to date anyway.

kasperg committed:
Mark all dependencies as dev dependencies

Regular dependencies will be added to projects which depend on this
package. However the current dependencies are not actually needed when
using the output of the package at runtime. They are only needed when
working with Storybook for development.

Consequently we can mark all the dependencies as dev dependencies.
This way dependers only get the built CSS, JS and other assets.

kasperg committed:
Do not mark Chromatic action as failed despite errors

GitHub Actions does not have an easy way to make an action 
conditional on the availability of a secret. This may be the simplest
way to get around this. 

Remote unnecessary comments while we are at it.

kasperg committed:
Add prefix to tags created for automatic releases

Otherwise we end up in situations where branch and tag names overlap.
Git does not like this and will complain about duplicate ref specs
when trying to push to the branch.

We cannot drop the tags entirely. GitHub requires a tag to create a
non-draft reelase.

kasperg committed:
Deploy on all pushes

It seems as if limiting to branches and tags will prevent the
workflow from running despite using wildcards. Perhaps it will only
trigger for creation of tags/branches.

Adamik10 committed:
Removal of engines setting in package.json

Package managers will not allow installation of packages which require a newer engine than what is available. ddb-react currently requires Node 12 which conflicts with the current package and will prevent installation. Even though this package only requires Node 16 for development NPM & Co has no way to distinguish between development and runtime requirements.

To resolve the situation we end up removing the engine specification from package.json. To compensate we add Node version requirements to .nvmrc so it will be picked up more easily here. Otherwise we just have to hope developers follow the README.

kasperg commented on package.json:
```diff
@@ -1,27 +1,8 @@
-    "typescript": "^4.1.2",
-    "web-vitals": "^1.0.1"
-  },
-  "engines": {
-    "node": ">=16",
-    "npm": ">=7"
-  },
```
The commit comment for this change is:

>  Test
> - removal of engines setting in package.json and adding .nvmrc
> - we are doing this in order to be compatible with the node version of ddb-react

Next time consider improving the headline and adding a few more details about your thought process here. A suggestion might be:

>  Remove of engines setting in package.json
>
> Package managers will not allow installation of packages which require a newer engine than what is available. ddb-react currently requires Node 12 which conflicts with the current package and will prevent installation. Even though this package only requires Node 16 for development NPM & Co has no way to distinguish between development and runtime requirements.
>
> To resolve the situation we end up removing the engine specification from package.json. To compensate we add Node version requirements to `.nvmrc` so it will be picked up more easily here. Otherwise we just have to hope developers follow the README.

Perhaps a bit wordy but the important parts are:

1. Getting in the headline right
2. Making clear the reasons why you made the change in the first place—the way things worked before the change (and what was wrong with that), the way they work now, and why you decided to solve it the way you did

https://cbea.ms/git-commit/#why-not-how is a great resource about commit messages :smile:

Adamik10 replied:
Thank you @kasperg, I amended this and I shall do better with commit messages in the future.

kasperg reviewed (approved):
:+1: but that almost goes without saying. The code is basicly what we have already agreed on in #1 😄 .

Adamik10 commented:
@spaceo @kasperg what are these "UI Tests"?
![image](https://user-images.githubusercontent.com/28546954/164684478-203ef731-92e0-43f9-9a1a-218c8008846b.png)

kasperg commented:
@Adamik10 These are Chromatic tests which will highlight UI changes based on the code changes.

Chromatic has a review flow with allows us to accept or reject changes which we may want to use or not.

kasperg commented on .github/workflows/deployment.yml:
```diff
@@ -29,12 +25,63 @@ jobs:
+        run: |
+          npm version 0.0.0-$GITHUB_SHA --no-git-tag-version
+          cat package.json
+          npm publish --tag $GITHUB_REF_NAME
```
Branch names may contain special characters such as `/` which will not work as tag names for NPM packages. We talked about using a GitHub Action for this but perhaps it can be solved using a little Bash instead?


Pseudo code:

```
$NPM_TAG_NAME = replace("/", "-", $GITHUB_REF_NAME)
npm publish --tag $NPM_TAG_NAME
```

kasperg replied:
Working code:

```bash
 NPM_TAG_NAME=$(echo $GITHUB_REF_NAME | tr -d [:space:] | tr -C [:alnum:] _)
 npm publish --tag $NPM_TAG_NAME
 ```

Adamik10 committed:
Prevented the npm package to include special characters in its tag name

Adjusted the deployment github action so that it cannot use special
characters for npm package tag name.

Before, we directly copied the branch name in order to make the tag.

Branch names often use special characters like slashes, which are not
allowed in npm tag names.
